cou/apps/app.py

```diff
@@ -17,25 +17,114 @@
 
 import logging
 from collections import defaultdict
+from collections.abc import Callable
 from dataclasses import dataclass, field
 from io import StringIO
 from typing import Any, Optional
 
 from juju.client._definitions import ApplicationStatus
 from ruamel.yaml import YAML
 
-from cou.exceptions import MismatchedOpenStackVersions
-from cou.utils.juju_utils import extract_charm_name_from_url
+from cou.exceptions import (
+    ApplicationError,
+    HaltUpgradePlanGeneration,
+    MismatchedOpenStackVersions,
+)
+from cou.steps import UpgradeStep
+from cou.utils.juju_utils import (
+    async_get_status,
+    async_set_application_config,
+    async_upgrade_charm,
+)
 from cou.utils.openstack import OpenStackCodenameLookup, OpenStackRelease
 
 logger = logging.getLogger(__name__)
 
 
+class AppFactory:
+    """Factory class for Application objects."""
+
+    apps_type: dict[str, type[OpenStackApplication]] = {}
+
+    @classmethod
+    def create(
+        cls,
+        name: str,
+        status: ApplicationStatus,
+        config: dict,
+        model_name: str,
+        charm: str,
+    ) -> Optional[OpenStackApplication]:
+        """Create the OpenStackApplication or registered subclasses.
+
+        Applications Subclasses registered with the "register_application"
+        decorator can be instantiated and used with their customized methods.
+        :param name: Name of the application
+        :type name: str
+        :param status: Status of the application
+        :type status: ApplicationStatus
+        :param config: Configuration of the application
+        :type config: dict
+        :param model_name: Model name
+        :type model_name: str
+        :param charm: Name of the charm
+        :type charm: str
+        :return: The OpenStackApplication class or None if not supported.
+        :rtype: Optional[OpenStackApplication]
+        """
+        # pylint: disable=too-many-arguments
+        if OpenStackCodenameLookup.charm_supported(charm):
+            app_class = cls.apps_type.get(name, OpenStackApplication)
+            if status.subordinate_to:
+                logger.warning(
+                    (
+                        "'%s' is a subordinate application and it's not currently "
+                        "supported for upgrading"
+                    ),
+                    name,
+                )
+                return None
+            return app_class(
+                name=name, status=status, config=config, model_name=model_name, charm=charm
+            )
+        logger.warning(
```

Comment:
warning may be too much. There may be several spurious apps in an openstack model.

Reply:
Agree. Changed to debug